### DISCUSSION TITLE
Is it possible to implement interoperability with C++ and Objective-C/C++?

### DISCUSSION THREAD

**Discussion by przemyslaw0** [edubart/nelua-lang#217] [Q&A] [answered]
Is it possible to implement interoperability with C++, Objective-C, and Objective-C++ by using Nelua metaprogramming facilities?

If some of the Nelua targets are system programming and making of libraries, interact directly with C++ and Objective-C/C++ would be a very good idea.

**Comment by asumbek** [comment, marked as answer]
Yes, it is possible. But requires you to write a wrapper for Nelua. Here's an [example of interoperability with C++ in Nelua](https://github.com/UrNightmaree/nelua-cpp-example).

As you can see, there's a wrapper for the Hello library (`hello.cpp`). It contains functions that wraps C++ functions using `extern "C"`. In `hello.nelua`, it contains several definition of `libhello` without `nodecl` attribute, and a code that'll run a function from `hello.cpp` using `libhello`.

It's basically like [FFI <-> C++ method](https://github.com/cslarsen/luajit-cpp), you make a wrapper of it, compile it to shared library, and then use FFI to define and use the functions. Except the difference is, you don't use `dlopen` to get the functions, since you can define the functions into your `.nelua`, and also you're not limited to only compile it to shared library, static library can also be used.

**Reply by przemyslaw0** [reply]
Just another question, is it possible to use non-C features like classes in the Nelua file? I know that Nelua has no OOP support, but is it possible to write a macro to use the classes of C++ file?

**Reply by asumbek** [reply]
Seems not, but you can try look at [an example](https://github.com/edubart/nelua-lang/raw/master/examples/record_inheretance.nelua) in the Nelua repository.
